File: copilot/backend/repositories/inference_scenario_binding_repository.py
"""Persistence for Copilot-owned Inference Scenario Bindings."""

from sqlalchemy import select
from sqlalchemy.orm import Session

from models.inference_scenario_binding import InferenceScenarioBinding


class InferenceScenarioBindingVersionConflict(RuntimeError):
    pass


class InferenceScenarioBindingRepository:
    def __init__(self, db: Session) -> None:
        self._db = db

    def resolve(self, tenant_id: int, scenario_code: str) -> InferenceScenarioBinding | None:
        tenant = self._db.scalar(
            select(InferenceScenarioBinding).where(
                InferenceScenarioBinding.scenario_code == scenario_code,
                InferenceScenarioBinding.scope_type == "tenant",
                InferenceScenarioBinding.tenant_id == tenant_id,
            )
        )
        if tenant is not None:
            return tenant
        return self._db.scalar(
            select(InferenceScenarioBinding).where(
                InferenceScenarioBinding.scenario_code == scenario_code,
                InferenceScenarioBinding.scope_type == "platform",
                InferenceScenarioBinding.tenant_id.is_(None),
            )
        )

    def get(self, scope_type: str, tenant_id: int | None, scenario_code: str) -> InferenceScenarioBinding | None:
        statement = select(InferenceScenarioBinding).where(
            InferenceScenarioBinding.scenario_code == scenario_code,
            InferenceScenarioBinding.scope_type == scope_type,
        )
        statement = statement.where(
            InferenceScenarioBinding.tenant_id.is_(None)
            if tenant_id is None
            else InferenceScenarioBinding.tenant_id == tenant_id
        )
        return self._db.scalar(statement)

    def save(
        self,
        *,
        scope_type: str,
        tenant_id: int | None,
        scenario_code: str,
        model_profile_id: str,
        expected_version: int,
        updated_by: int,
    ) -> InferenceScenarioBinding:
        statement = select(InferenceScenarioBinding).where(
            InferenceScenarioBinding.scenario_code == scenario_code,
            InferenceScenarioBinding.scope_type == scope_type,
        )
        statement = statement.where(
            InferenceScenarioBinding.tenant_id.is_(None)
            if tenant_id is None
            else InferenceScenarioBinding.tenant_id == tenant_id
        ).with_for_update()
        current = self._db.scalar(statement)
        if current is None:
            if expected_version != 0:
                raise InferenceScenarioBindingVersionConflict()
            current = InferenceScenarioBinding(
                scenario_code=scenario_code,
                scope_type=scope_type,
                tenant_id=tenant_id,
                model_profile_id=model_profile_id,
                version=1,
                updated_by=updated_by,
            )
            self._db.add(current)
        else:
            if current.version != expected_version:
                raise InferenceScenarioBindingVersionConflict()
            current.model_profile_id = model_profile_id
            current.version += 1
            current.updated_by = updated_by
        self._db.commit()
        self._db.refresh(current)
        return current

File: manager/backend/internal/repository/inference_scenario_binding_repository.go
package repository

import (
	"context"
	"errors"

	"github.com/addp/manager/internal/models"
	"gorm.io/gorm"
	"gorm.io/gorm/clause"
)

var ErrInferenceScenarioBindingVersionConflict = errors.New("inference scenario binding version conflict")

type InferenceScenarioBindingRepository struct {
	db *gorm.DB
}

func NewInferenceScenarioBindingRepository(db *gorm.DB) *InferenceScenarioBindingRepository {
	return &InferenceScenarioBindingRepository{db: db}
}

func (r *InferenceScenarioBindingRepository) Resolve(ctx context.Context, tenantID uint, scenario string) (*models.InferenceScenarioBinding, error) {
	var value models.InferenceScenarioBinding
	err := r.db.WithContext(ctx).
		Where("scenario_code = ? AND scope_type = 'tenant' AND tenant_id = ?", scenario, tenantID).
		First(&value).Error
	if err == nil {
		return &value, nil
	}
	if !errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, err
	}
	err = r.db.WithContext(ctx).
		Where("scenario_code = ? AND scope_type = 'platform' AND tenant_id IS NULL", scenario).
		First(&value).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &value, err
}

func (r *InferenceScenarioBindingRepository) Get(ctx context.Context, scopeType string, tenantID *uint, scenario string) (*models.InferenceScenarioBinding, error) {
	query := r.db.WithContext(ctx).Where("scenario_code = ? AND scope_type = ?", scenario, scopeType)
	if tenantID == nil {
		query = query.Where("tenant_id IS NULL")
	} else {
		query = query.Where("tenant_id = ?", *tenantID)
	}
	var value models.InferenceScenarioBinding
	err := query.First(&value).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &value, err
}

func (r *InferenceScenarioBindingRepository) Save(ctx context.Context, value *models.InferenceScenarioBinding, expectedVersion uint64) error {
	return r.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		query := tx.Clauses(clause.Locking{Strength: "UPDATE"}).
			Where("scenario_code = ? AND scope_type = ?", value.ScenarioCode, value.ScopeType)
		if value.TenantID == nil {
			query = query.Where("tenant_id IS NULL")
		} else {
			query = query.Where("tenant_id = ?", *value.TenantID)
		}
		var current models.InferenceScenarioBinding
		err := query.First(&current).Error
		if errors.Is(err, gorm.ErrRecordNotFound) {
			if expectedVersion != 0 {
				return ErrInferenceScenarioBindingVersionConflict
			}
			value.Version = 1
			return tx.Create(value).Error
		}
		if err != nil {
			return err
		}
		if current.Version != expectedVersion {
			return ErrInferenceScenarioBindingVersionConflict
		}
		value.ID = current.ID
		value.Version = current.Version + 1
		result := tx.Model(&models.InferenceScenarioBinding{}).
			Where("id = ? AND version = ?", current.ID, expectedVersion).
			Updates(map[string]interface{}{
				"model_profile_id": value.ModelProfileID,
				"version":          value.Version,
				"updated_by":       value.UpdatedBy,
			})
		if result.Error != nil {
			return result.Error
		}
		if result.RowsAffected != 1 {
			return ErrInferenceScenarioBindingVersionConflict
		}
		return tx.First(value, current.ID).Error
	})
}
